### 1.29异常

Java 异常类层次结构图

![Java异常类层次结构图](https://my-blog-to-use.oss-cn-beijing.aliyuncs.com/2019-2/Exception.png)

java中，所有的异常都有一个共同的祖先，java.lang包中的**Throwable类**；

Throwable：有两个重要的子类：**Error（错误）**和 **Exception（异常）**，二者都是java异常处理的重要子类，各自都包含大量子类；

****

**Error(错误)：是程序无法处理的错误**，表示运行应用程序中比较严重问题。大多数错误与代码编写者执行的操作无关，而表示代码运行时JVM（java虚拟机）出现的问题。

例如：java虚拟机运行错误（Virtual MachineError）,当JVM不再有继续执行操作所需的内存资源时，将出现OutOfMemoryError。这些异常发生时，java虚拟机（JVM）一般会选择线程终止。

这些错误表示故障发生于虚拟机自身，或者发生在虚拟机试图执行应用时，如java虚拟机运行错误（Virtual MachineError），类定义错误（NoClassDefFoundError）等。这些错误是不可查的，因为它们在应用程序的控制和处理能力之外，而且绝大多数是程序运行时不允许出现的状况。对于设计合理的应用程序来说，即使确定发生了错误，本质上也不应该试图去处理它所引起的异常状况，在java中，错误通过Error的子类描述；

****

**Exception(异常):是程序本身可以处理的异常**。Exception类有一个重要的子类**RuntimeExcpetion**它是由java虚拟机抛出。

**NullPointerException**(要访问的变量没有引用任何对象，抛出该异常)；

**ArithmeticException**(算术运算异常，一个整数除以0时，抛出该异常)；

**ArrayIndexOutOfBoundsException** (下标越界异常)；

注意：**异常和错误的区别：异常能被程序本身处理，错误时无法处理的；**

****

**Throwable类常用方法**

- public String getMessage():返回异常发生时的简要描述；
- public String toString():返回异常发生时的详细信息；
- public String getLocalizedMessage():返回异常对象的本地化信息，使用Throwable的子类覆盖这个方法，可以生成本地化信息，如果子类没有覆盖该方法，则该方法返回的信息与getMessage()返回的结果相同；
- public void printStrackTrace():在控制台上打印Throwable对象封装的异常信息；

****

**异常处理总结**

- try块：用于捕获异常，其后可接零个或多个catch块，如果没有catch块，则必须跟一个finally块；
- catch块：用于处理try捕获到的异常；
- finally块：无论是否捕获或处理异常，finally块里的语句都会被执行。当在try块或catch块中遇到return语句时，finally语句将在方法返回之前被执行；

**在以下4种特殊情况下，finally块不会被执行**

1. 在finally语句第一行发生异常，因为在其他行，finally块还是会得到执行；
2. 在前面的代码中用了System.exit(int)已退出程序，exit时带参函数；若该语句在异常语句之后，finally会执行；
3. 程序所在的线程死亡；
4. 关闭CPU